File: test_client.py
#!/usr/bin/env python3
"""
Тестовый клиент для проверки OCR сервера
Использование:
    python test_client.py --host 192.168.1.10 --port 8080 --image test.jpg
"""

import argparse
import base64
import json
import requests
import websocket
import sys
from pathlib import Path


def test_http_api(host, port, image_path, language=None):
    """Тестирование HTTP API"""
    url = f"http://{host}:{port}/ocr"
    
    print(f"Тестирование HTTP API: {url}")
    if language:
        print(f"Язык: {language}")
    
    with open(image_path, 'rb') as f:
        files = {'image': f}
        data = {'language': language} if language else {}
        response = requests.post(url, files=files, data=data)
    
    print(f"Статус код: {response.status_code}")
    
    if response.status_code == 200:
        result = response.json()
        print(f"Успешно! Распознанный текст:")
        print(f"  Текст: {result.get('text', 'N/A')}")
        print(f"  Язык: {result.get('language', 'N/A')}")
        print(f"  Уверенность: {result.get('confidence', 'N/A')}")
        print(f"  Время обработки: {result.get('processingTimeMs', 'N/A')}ms")
    else:
        print(f"Ошибка: {response.text}")


def test_websocket(host, port, image_path, language=None):
    """Тестирование WebSocket API"""
    ws_url = f"ws://{host}:{port}/ws"
    
    print(f"\nТестирование WebSocket API: {ws_url}")
    if language:
        print(f"Язык: {language}")
    
    with open(image_path, 'rb') as f:
        image_data = base64.b64encode(f.read()).decode('utf-8')
    
    def on_message(ws, message):
        data = json.loads(message)
        msg_type = data.get('type', 'unknown')
        
        print(f"[{msg_type}] ", end='')
        
        if msg_type == 'status':
            print(f"{data.get('message', 'N/A')}")
        elif msg_type == 'progress':
            print(f"Прогресс: {data.get('progress', 0)}%")
        elif msg_type == 'result':
            print(f"Результат получен!")
            print(f"  Текст: {data.get('text', 'N/A')}")
            print(f"  Язык: {data.get('language', 'N/A')}")
            print(f"  Уверенность: {data.get('confidence', 'N/A')}")
            print(f"  Время обработки: {data.get('processingTimeMs', 'N/A')}ms")
            ws.close()
        elif msg_type == 'error':
            print(f"Ошибка: {data.get('error', 'N/A')}")
            ws.close()
        else:
            print(json.dumps(data, ensure_ascii=False))
    
    def on_error(ws, error):
        print(f"Ошибка WebSocket: {error}")
    
    def on_close(ws, close_status_code, close_msg):
        print("WebSocket соединение закрыто")
    
    def on_open(ws):
        print("WebSocket соединение установлено")
        message = {
            'type': 'ocr',
            'image': f'data:image/jpeg;base64,{image_data}'
        }
        if language:
            message['language'] = language
        ws.send(json.dumps(message))
    
    ws = websocket.WebSocketApp(
        ws_url,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close
    )
    
    ws.run_forever()


def test_status(host, port):
    """Проверка статуса сервера"""
    url = f"http://{host}:{port}/status"
    
    print(f"\nПроверка статуса сервера: {url}")
    
    try:
        response = requests.get(url)
        if response.status_code == 200:
            result = response.json()
            print("Статус сервера:")
            print(f"  Состояние: {result.get('status', 'N/A')}")
            print(f"  Uptime: {result.get('uptime', 0) / 1000:.1f} секунд")
            print(f"  Количество запросов: {result.get('requestCount', 0)}")
            print(f"  Порт: {result.get('port', 'N/A')}")
            print(f"  OCR движок: {result.get('ocrEngine', 'N/A')}")
        else:
            print(f"Ошибка: {response.status_code} - {response.text}")
    except Exception as e:
        print(f"Не удалось подключиться к серверу: {e}")


def main():
    parser = argparse.ArgumentParser(
        description='Тестовый клиент для OCR сервера'
    )
    parser.add_argument(
        '--host',
        default='192.168.1.10',
        help='IP адрес сервера (по умолчанию: 192.168.1.10)'
    )
    parser.add_argument(
        '--port',
        type=int,
        default=8080,
        help='Порт сервера (по умолчанию: 8080)'
    )
    parser.add_argument(
        '--image',
        help='Путь к изображению для распознавания'
    )
    parser.add_argument(
        '--language',
        help='Язык для OCR (eng - английский, rus - русский, auto - автоопределение)'
    )
    parser.add_argument(
        '--mode',
        choices=['http', 'ws', 'both', 'status'],
        default='status',
        help='Режим тестирования (по умолчанию: status)'
    )
    
    args = parser.parse_args()
    
    if args.mode == 'status':
        test_status(args.host, args.port)
    elif args.mode in ['http', 'ws', 'both']:
        if not args.image:
            print("Ошибка: необходимо указать путь к изображению (--image)")
            sys.exit(1)
        
        if not Path(args.image).exists():
            print(f"Ошибка: файл {args.image} не найден")
            sys.exit(1)
        
        if args.mode in ['http', 'both']:
            test_http_api(args.host, args.port, args.image, args.language)
        
        if args.mode in ['ws', 'both']:
            test_websocket(args.host, args.port, args.image, args.language)


if __name__ == '__main__':
    main()

File: test_client.js
#!/usr/bin/env node
/**
 * Тестовый клиент для проверки OCR сервера (JavaScript/Node.js)
 * 
 * Установка зависимостей:
 *   npm install ws axios form-data
 * 
 * Использование:
 *   node test_client.js --host 192.168.1.10 --port 8080 --image test.jpg --mode status
 */

const fs = require('fs');
const axios = require('axios');
const WebSocket = require('ws');
const FormData = require('form-data');

function parseArgs() {
    const args = {
        host: '192.168.1.10',
        port: 8080,
        image: null,
        language: null,
        mode: 'status'
    };
    
    for (let i = 2; i < process.argv.length; i++) {
        if (process.argv[i] === '--host' && process.argv[i + 1]) {
            args.host = process.argv[++i];
        } else if (process.argv[i] === '--port' && process.argv[i + 1]) {
            args.port = parseInt(process.argv[++i]);
        } else if (process.argv[i] === '--image' && process.argv[i + 1]) {
            args.image = process.argv[++i];
        } else if (process.argv[i] === '--language' && process.argv[i + 1]) {
            args.language = process.argv[++i];
        } else if (process.argv[i] === '--mode' && process.argv[i + 1]) {
            args.mode = process.argv[++i];
        }
    }
    
    return args;
}

async function testHttpApi(host, port, imagePath, language = null) {
    const url = `http://${host}:${port}/ocr`;
    
    console.log(`Тестирование HTTP API: ${url}`);
    if (language) {
        console.log(`Язык: ${language}`);
    }
    
    try {
        const form = new FormData();
        form.append('image', fs.createReadStream(imagePath));
        if (language) {
            form.append('language', language);
        }
        
        const response = await axios.post(url, form, {
            headers: form.getHeaders()
        });
        
        console.log(`Статус код: ${response.status}`);
        console.log('Успешно! Распознанный текст:');
        console.log(`  Текст: ${response.data.text || 'N/A'}`);
        console.log(`  Язык: ${response.data.language || 'N/A'}`);
        console.log(`  Уверенность: ${response.data.confidence || 'N/A'}`);
        console.log(`  Время обработки: ${response.data.processingTimeMs || 'N/A'}ms`);
    } catch (error) {
        console.error(`Ошибка: ${error.message}`);
        if (error.response) {
            console.error(`Статус: ${error.response.status}`);
            console.error(`Данные: ${JSON.stringify(error.response.data)}`);
        }
    }
}

function testWebSocket(host, port, imagePath, language = null) {
    const wsUrl = `ws://${host}:${port}/ws`;
    
    console.log(`\nТестирование WebSocket API: ${wsUrl}`);
    if (language) {
        console.log(`Язык: ${language}`);
    }
    
    const imageData = fs.readFileSync(imagePath, { encoding: 'base64' });
    const ws = new WebSocket(wsUrl);
    
    ws.on('open', () => {
        console.log('WebSocket соединение установлено');
        
        const messageObj = {
            type: 'ocr',
            image: `data:image/jpeg;base64,${imageData}`
        };
        if (language) {
            messageObj.language = language;
        }
        const message = JSON.stringify(messageObj);
        
        ws.send(message);
    });
    
    ws.on('message', (data) => {
        const message = JSON.parse(data.toString());
        const msgType = message.type || 'unknown';
        
        process.stdout.write(`[${msgType}] `);
        
        switch (msgType) {
            case 'status':
                console.log(message.message || 'N/A');
                break;
            case 'progress':
                console.log(`Прогресс: ${message.progress || 0}%`);
                break;
            case 'result':
                console.log('Результат получен!');
                console.log(`  Текст: ${message.text || 'N/A'}`);
                console.log(`  Язык: ${message.language || 'N/A'}`);
                console.log(`  Уверенность: ${message.confidence || 'N/A'}`);
                console.log(`  Время обработки: ${message.processingTimeMs || 'N/A'}ms`);
                ws.close();
                break;
            case 'error':
                console.log(`Ошибка: ${message.error || 'N/A'}`);
                ws.close();
                break;
            default:
                console.log(JSON.stringify(message));
        }
    });
    
    ws.on('error', (error) => {
        console.error(`Ошибка WebSocket: ${error.message}`);
    });
    
    ws.on('close', () => {
        console.log('WebSocket соединение закрыто');
    });
}

async function testStatus(host, port) {
    const url = `http://${host}:${port}/status`;
    
    console.log(`\nПроверка статуса сервера: ${url}`);
    
    try {
        const response = await axios.get(url);
        
        console.log('Статус сервера:');
        console.log(`  Состояние: ${response.data.status || 'N/A'}`);
        console.log(`  Uptime: ${(response.data.uptime || 0) / 1000} секунд`);
        console.log(`  Количество запросов: ${response.data.requestCount || 0}`);
        console.log(`  Порт: ${response.data.port || 'N/A'}`);
        console.log(`  OCR движок: ${response.data.ocrEngine || 'N/A'}`);
    } catch (error) {
        console.error(`Не удалось подключиться к серверу: ${error.message}`);
    }
}

async function main() {
    const args = parseArgs();
    
    if (args.mode === 'status') {
        await testStatus(args.host, args.port);
    } else if (['http', 'ws', 'both'].includes(args.mode)) {
        if (!args.image) {
            console.error('Ошибка: необходимо указать путь к изображению (--image)');
            process.exit(1);
        }
        
        if (!fs.existsSync(args.image)) {
            console.error(`Ошибка: файл ${args.image} не найден`);
            process.exit(1);
        }
        
        if (args.mode === 'http' || args.mode === 'both') {
            await testHttpApi(args.host, args.port, args.image, args.language);
        }
        
        if (args.mode === 'ws' || args.mode === 'both') {
            testWebSocket(args.host, args.port, args.image, args.language);
        }
    } else {
        console.error('Неизвестный режим. Используйте: status, http, ws, или both');
        process.exit(1);
    }
}

main();

